File: Data Structures & Algorithms/balanced-binary-tree/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:

    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        return self.dfs(root) != -1

    def dfs(self, root: Optional[TreeNode]) -> int:
        # base case
        if not root:
            return 0
        
        leftDepth = self.dfs(root.left)
        rightDepth = self.dfs(root.right)

        if leftDepth == -1 or rightDepth == -1 or abs(leftDepth - rightDepth) > 1:
            return -1

        return max(leftDepth, rightDepth) + 1

File: Data Structures & Algorithms/binary-tree-diameter/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
private:
    int diameter = 0;
public:

    int diameterOfBinaryTree(TreeNode* root) {
        dfs(root);
        return diameter;
    }

    int dfs(TreeNode* root) {
        // base case
        if (!root) {
            return 0;
        }
        
        int leftDepth = dfs(root->left);
        int rightDepth = dfs(root->right);

        diameter = max(leftDepth + rightDepth, diameter);
        return max(leftDepth, rightDepth) + 1;
    }
};
